File: app/src/main/java/closer/vlllage/com/closer/ui/TimedValue.java
package closer.vlllage.com.closer.ui;

import java.util.Date;
import java.util.LinkedList;

public class TimedValue<T> {
    private final int trackMsInPast;
    private LinkedList<Capture<T>> trackedValues = new LinkedList<>();

    public TimedValue(int trackMsInPast) {
        this.trackMsInPast = trackMsInPast;
    }

    public void report(T value) {
        trackedValues.add(new Capture<>(value, new Date()));
        cleanup();
    }

    public T get() {
        cleanup();
        return trackedValues.isEmpty() ? null : trackedValues.getFirst().value;
    }

    public T now() {
        cleanup();
        return trackedValues.isEmpty() ? null : trackedValues.getLast().value;
    }

    private void cleanup() {
        Date now = new Date();
        now.setTime(now.getTime() - trackMsInPast);
        while (!trackedValues.isEmpty()) {
            if (!trackedValues.getFirst().time.before(now)) {
                break;
            }
            trackedValues.removeFirst();
        }
    }

    private static class Capture<T> {
        final T value;
        final Date time;

        public Capture(T value, Date time) {
            this.value = value;
            this.time = time;
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/ui/TimedValue.kt
package closer.vlllage.com.closer.ui

import java.util.*

class TimedValue<T>(private val trackMsInPast: Int) {
    private val trackedValues = LinkedList<Capture<T>>()

    fun report(value: T) {
        trackedValues.add(Capture(value, Date()))
        cleanup()
    }

    fun get(): T? {
        cleanup()
        return if (trackedValues.isEmpty()) null else trackedValues.first.value
    }

    fun now(): T? {
        cleanup()
        return if (trackedValues.isEmpty()) null else trackedValues.last.value
    }

    private fun cleanup() {
        val now = Date()
        now.time = now.time - trackMsInPast
        while (!trackedValues.isEmpty()) {
            if (!trackedValues.first.time.before(now)) {
                break
            }
            trackedValues.removeFirst()
        }
    }

    private class Capture<T>(internal val value: T, internal val time: Date)
}
